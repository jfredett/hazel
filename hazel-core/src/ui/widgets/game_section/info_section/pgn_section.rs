use ratatui::layout::Direction;
use ratatui::prelude::*;

use crate::notation::pgn::PGN;
use crate::ui::widgets::placeholder::Placeholder;

lazy_static! {
    static ref LAYOUT : Layout = Layout::default()
        .direction(Direction::Horizontal)
        .constraints(
            [
                Constraint::Percentage(51),
                Constraint::Percentage(49)
            ].as_ref(),
        );
}


// this should be a stateful widget, it's struct-state should just be style information
// information, the provided state will be the current position of the cursor and what to display
// in the box.
//
// The parent will hold t

#[derive(Default)]
pub struct PGNSection {
    current_position: usize,
    pgn: PGN,
}

impl PGNSection {
    pub fn new(pgn: PGN) -> Self {
        Self { current_position: 0, pgn }
    }

    pub fn set_position(&mut self, position: usize) {
        self.current_position = position;
    }

    pub fn advance(&mut self) {
        self.current_position += 1;
    }

    pub fn retreat(&mut self) {
        self.current_position -= 1;
    }

    pub fn render_variation(&self, area: Rect, buf: &mut Buffer) {
        let mut pgn = self.pgn.clone();
        let mut s = vec![]; 
        let mut f = pgn.familiar();

        f.advance_until(|f_inner| {

            if !f_inner.move_string().is_empty() {
                s.push(f_inner.move_string());
            }

            f_inner.cursor_position() == self.current_position
        });

        let text = s.chunks(2).enumerate().map(|(idx, chunk)| {
            format!("{:2}. {}", idx + 1, chunk.join(" "))
        }).collect::<Vec<String>>().join("\n");

        Text::styled(text, Style::default()).render(area, buf);
    }
}

impl Widget for &PGNSection {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let chunks = LAYOUT.split(area);

        self.render_variation(chunks[0], buf);

        let query = Placeholder::of_size(chunks[1].width, chunks[1].height);
        Widget::render(&query, chunks[1], buf);
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::notation::pgn::PGN;

    use super::*;

    fn example_game() -> PGN {
        PGN::load("tests/fixtures/no-variations-and-halts.pgn").unwrap()
    }

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 16);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let board_section = &mut PGNSection::new(example_game());
        board_section.set_position(16);
        board_section.render(rect, &mut buffer);

        assert_debug_snapshot!(buffer)
    }
}

