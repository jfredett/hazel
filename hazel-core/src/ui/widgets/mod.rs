// Sections are direct representations of chunks of the UI
pub mod tile;

pub mod engine_io_section;
pub mod game_section;

// other widgets are more general use oriented.
pub mod query;
pub mod placeholder;
pub mod tapereader;

pub mod input;
pub mod output;

pub mod board;
pub mod smallboard;
pub mod fen;

