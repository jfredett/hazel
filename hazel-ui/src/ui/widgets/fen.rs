use ratatui::{buffer::Buffer, prelude::*, text::Text, widgets::Widget};

use crate::notation::ben::BEN;


#[derive(Debug, Default)]
#[allow(clippy::upper_case_acronyms)]
pub struct FEN {
    fen: BEN,
    style: Style,
    alignment: Alignment
}

impl FEN {
    pub fn new(fen: BEN) -> Self {
        Self {
            fen,
            style: Style::default(),
            alignment: Alignment::Left
        }
    }

    pub fn set_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }

    pub fn alignment(mut self, alignment: Alignment) -> Self {
        self.alignment = alignment;
        self
    }

    #[inline(always)]
    pub fn center(self) -> Self {
        self.alignment(Alignment::Center)
    }
}

impl Widget for &FEN {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let fenstring = format!("{}", self.fen);
        Text::styled(fenstring, self.style).alignment(self.alignment).render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use crate::constants::START_POSITION_FEN;
    use crate::board::simple::PieceBoard;

    use super::*;

    #[test]
    fn renders_empty_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let fen_widget = &FEN::new(BEN::empty());
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "8/8/8/8/8/8/8/8 w KQkq - 0 1                                    "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn render_startpos_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let fen_widget = &FEN::new(BEN::new(START_POSITION_FEN));
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1        "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);

    }
}
