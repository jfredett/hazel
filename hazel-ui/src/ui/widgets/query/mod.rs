// THis is a placeholder, eventually it will support arbitary queries against the Hazel UI backend
// (and thus the engines it's managing) via some custom repl langauge
