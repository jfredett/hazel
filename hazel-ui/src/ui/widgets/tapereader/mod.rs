use ratatui::{buffer::Buffer, layout::{Constraint, Layout, Rect}, style::{Style, Stylize}, widgets::{Block, StatefulWidget, Table, Widget}};
use hazel::types::tape::familiar::state::tape_reader_state::TapeReaderState;

#[derive(Default)]
pub struct TapeReaderWidget {
    pub desired_position: usize
}

impl TapeReaderWidget {
    pub fn layout(&self) -> Layout {
        Layout::vertical([
            Constraint::Length(1), // header
            Constraint::Length(40), // code seciton
            Constraint::Min(1), // footer
        ])
    }

    pub fn advance(&mut self) {
        tracing::trace!(target="hazel::ui::events", "advancing to desired position: {:#04X} from {:#04X}", self.desired_position + 1, self.desired_position);
        self.desired_position += 1;
    }

    pub fn rewind(&mut self) {
        tracing::trace!(target="hazel::ui::events", "rewinding to desired position: {:#04X} from {:#04X}", self.desired_position.saturating_sub(1), self.desired_position);
        self.desired_position = self.desired_position.saturating_sub(1);
    }
}


impl StatefulWidget for &TapeReaderWidget {
    type State = TapeReaderState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        // Columns widths are constrained in the same way as Layout...

        let chunks = self.layout().split(area);
        let header = chunks[0];
        let tape_area = chunks[1];
        let footer = chunks[2];

        let widths = [
            Constraint::Length(8),
            Constraint::Length(32),
            Constraint::Length(32)
        ];

        // TODO: This is probably not identity, but some function of height
        // state.set_page_size((code.height - 8) as usize);

        let table = Table::new(state.rows(), widths)
            .column_spacing(1)
            .style(Style::new().white())
            .block(Block::bordered())
            .header(state.header_row())
            .row_highlight_style(Style::new().reversed())
            .column_highlight_style(Style::new().red())
            .cell_highlight_style(Style::new().blue())
            .highlight_symbol(">>");

        Widget::render(state.header(), header, buf);
        StatefulWidget::render(&table, tape_area, buf, &mut state.table_state());
        Widget::render(state.footer(), footer, buf);
    }
}

#[cfg(test)]
mod tests {
    // use insta::assert_debug_snapshot;
    // use ratatui::{buffer::Buffer, style::Color};

    // use hazel::{game::position::Position, notation::pgn::PGN};

    // use super::*;

    // // FIXME: familiar refactor
    // #[test]
    // fn renders_as_expected() {
    //     let rect = Rect::new(0, 0, 64, 32);
    //     let mut actual = Buffer::empty(rect);
    //     actual.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

    //     let pgn = PGN::load("tests/fixtures/no-variations-and-halts.pgn").unwrap();
    //     let mut position = Position::new(pgn.current_position());
    //     let tape = position.tape.read().unwrap();
    //     let mut fam = tape.conjure::<TapeReaderState>();

    //     let tapereader = TapeReaderWidget::default();

    //     tapereader.render(rect, &mut actual, fam.get_mut());

    //     assert_debug_snapshot!(actual);
    // }
}
