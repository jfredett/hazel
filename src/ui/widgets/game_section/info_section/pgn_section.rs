use ratatui::layout::Direction;
use ratatui::prelude::*;

use crate::ui::widgets::placeholder::Placeholder;

lazy_static! {
    static ref LAYOUT : Layout = Layout::default()
        .direction(Direction::Horizontal)
        .constraints(
            [
                Constraint::Percentage(51),
                Constraint::Percentage(49)
            ].as_ref(),
        );
}

pub struct PGNSection {
}

impl PGNSection {
    pub fn new() -> Self {
        Self {
        }
    }
}

impl StatefulWidget for &PGNSection {
    type State = ();
    fn render(self, area: Rect, buf: &mut Buffer, _state: &mut Self::State) {
        let chunks = LAYOUT.split(area);

        let pgn = Placeholder::of_size(chunks[0].width, chunks[0].height); // &mut pgnwidget::new();
        pgn.render(chunks[0], buf);//, state);

        // This is shown as time-per-move in the sketch, but should be swappable for whatever I
        // like.
        let query = Placeholder::of_size(chunks[1].width, chunks[1].height);
        query.render(chunks[1], buf);//, state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 16);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let board_section = &mut PGNSection::new();
        board_section.render(rect, &mut buffer, &mut ());

        let mut expected = Buffer::with_lines(vec![
            "┌───────────────────────────────┐┌─────────────────────────────┐",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "└───────────────────────────────┘└─────────────────────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }
}

