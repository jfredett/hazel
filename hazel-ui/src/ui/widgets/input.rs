use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

#[derive(Default)]
pub struct Input {
    content: String
}

impl Input {
    pub fn push(&mut self, input: char) {
        self.content.push(input);
    }

    pub fn pop(&mut self) {
        self.content.pop();
    }

    pub fn content(&self) -> String {
        self.content.clone()
    }

    pub fn flush(&mut self) -> String {
        let content = self.content.clone();
        self.content.clear();
        content
    }
}

impl StatefulWidget for &Input {
    type State = String;

    fn render(self, area: Rect, buf: &mut Buffer, _state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::NONE)
            .title(format!("$> {}", self.content))
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 3);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let input = &mut Input::default();
        input.push('a');
        input.push('b');
        input.push('c');
        input.render(rect, &mut buffer, &mut input.content());

        let mut expected = Buffer::with_lines(vec![
            "$> abc                                                          ",
            "                                                                ",
            "                                                                ",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn flush_emptys_input_buffer() {
        let mut input = Input::default();
        input.push('a');
        input.push('b');
        input.push('c');
        assert_eq!(input.flush(), "abc");
        assert_eq!(input.content(), "");
    }
}
