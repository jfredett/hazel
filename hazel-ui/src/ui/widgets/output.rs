
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

#[derive(Default)]
pub struct Output {
    buffer: Vec<String>
}

impl Output {
    pub fn push(&mut self, line: String) {
        self.buffer.push(line);
    }

    pub fn buffer(&self) -> Vec<String> {
        self.buffer.clone()
    }
}

fn wrap_string(s: &str, width: usize) -> Vec<String> {
    let mut lines = vec![];
    let mut working_copy = s.to_string();
    while working_copy.len() > width {
        let new_line = working_copy.split_off(width);
        lines.push(working_copy);
        working_copy = new_line;
    }
    lines.push(working_copy);
    lines
}

impl StatefulWidget for &Output {
    type State = Vec<String>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::ALL)
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);

        // long lines are wrapped and indented by 2 spaces
        let output = state.clone();
        let mut adjusted = vec![];
        for line in output.iter() {
            if line.len() > area.width as usize - 2 {
                let new_lines = wrap_string(line, area.width as usize - 2);
                adjusted.extend(new_lines.into_iter());
            } else {
                adjusted.push(line.to_string());
            }
        }

        // should fill from the bottom to the top:
        let mut y = area.bottom() - 2;
        for line in adjusted.into_iter().rev() {
            buf.set_string(area.left() + 1, y, line, Style::default().fg(Color::White).bg(Color::Black));
            y -= 1;
        }
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 3);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut Output::default();
        output.push("This is a test".to_string());
        output.render(rect, &mut buffer, &mut output.buffer().clone());

        let mut expected = Buffer::with_lines(vec![
            "┌──────────────────────────────────────────────────────────────┐",
            "│This is a test                                                │",
            "└──────────────────────────────────────────────────────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn renders_a_very_long_line_by_wrapping() {
        let rect = Rect::new(0, 0, 15, 6);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut Output::default();
        output.push("This is a very long line that should be wrapped".to_string());
        output.render(rect, &mut buffer, &mut output.buffer().clone());

        let mut expected = Buffer::with_lines(vec![
        "┌─────────────┐",
        "│This is a ver│",
        "│y long line t│",
        "│hat should be│",
        "│ wrapped     │",
        "└─────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }
}







