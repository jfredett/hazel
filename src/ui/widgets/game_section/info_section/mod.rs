use ratatui::layout::Direction;
use ratatui::prelude::*;
use ratatui::widgets::Borders;

use crate::ui::widgets::placeholder::Placeholder;

mod pgn_section;
use pgn_section::PGNSection;

lazy_static! {
    static ref LAYOUT : Layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(
            [
                Constraint::Max(3),
                Constraint::Fill(1)
            ].as_ref(),
        );
}

pub struct InfoSection {
}

impl InfoSection {
    pub fn new() -> Self {
        Self {
        }
    }
}

impl StatefulWidget for &InfoSection {
    type State = ();
    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let chunks = LAYOUT.split(area);

        let ticker_widget = Placeholder::of_size(chunks[0].width, chunks[0].height).borders(Borders::NONE); // &mut info_section::InfoSection::new();
        ticker_widget.render(chunks[0], buf);//, state);

        let pgn_section = &mut PGNSection::new();
        pgn_section.render(chunks[1], buf, state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 16);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let board_section = &mut InfoSection::new();
        board_section.render(rect, &mut buffer, &mut ());

        let mut expected = Buffer::with_lines(vec![
            "                           Placeholder                          ",
            "                           Placeholder                          ",
            "                           Placeholder                          ",
            "┌───────────────────────────────┐┌─────────────────────────────┐",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "│          Placeholder          ││         Placeholder         │",
            "└───────────────────────────────┘└─────────────────────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }
}

