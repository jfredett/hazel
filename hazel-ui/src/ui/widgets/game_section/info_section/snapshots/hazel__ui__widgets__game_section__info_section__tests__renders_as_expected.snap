---
source: src/ui/widgets/game_section/info_section/mod.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 64, height: 16 },
    content: [
        "                           Placeholder                          ",
        "                           Placeholder                          ",
        "                           Placeholder                          ",
        "                                 ┌─────────────────────────────┐",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 │         Placeholder         │",
        "                                 └─────────────────────────────┘",
    ],
    styles: [
        x: 0, y: 0, fg: White, bg: Black, underline: Reset, modifier: NONE,
    ]
}
